exposing resourcepack loading to api

onebeastchris opened the pull request:
hey!
This PR aims to bring an event that could be used by extensions (e.g.: [my extension for per-player-packs](https://github.com/onebeastchris/packs-extension/blob/c177ef9e5ea637ab901a566e9c96f97bc52da01a/src/main/java/net/onebeastchris/geyser/extension/packing/packing.java#L55) to send additional resource packs to a bedrock player joining via a Geyser instance.

Changes:
- Added ~~PlayerResourcePackLoadEvent~~ SessionResourcePackLoadEvent: method to add/get resourcepacks
- exposed ResourcePack & ResourcePackManifest classes to api; but left the resource pack loader in Geyser itself 
-> ~~since ResourcePackManifest has a few json annotations, i had to add an implementation dependency to the geyser api - I'm unsure if those annotations can/should be removed, or otherwise handled~~ resolved with interface split

Closes https://github.com/GeyserMC/Geyser/issues/3692 by adding a system property to hide the warning.

I'm open for changes/feedback/criticism - this is my first PR regarding any sort of API :)

Camotoy commented on api/src/main/java/org/geysermc/geyser/api/packs/ResourcePack.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * This represents a resource pack and all the data relevant to it
+ */
+@Data
```
I don't want to use Lombok in the API at all.

Redned235 commented on api/src/main/java/org/geysermc/geyser/api/packs/ResourcePack.java:
```diff
@@ -0,0 +1,42 @@
+@Data
+public class ResourcePack {
+    private byte[] sha256;
+    private File file;
```
```suggestion
    private Path path;
```

Everything in the API should use `java.nio` where possible for I/O.

Redned235 commented on core/src/main/java/org/geysermc/geyser/GeyserImpl.java:
```diff
@@ -258,7 +258,7 @@ private void startInstance() {
         SkinProvider.registerCacheImageTask(this);
 
-        ResourcePack.loadPacks();
+        ResourcePackUtil.loadPacks();
```
Would suggest moving resource packs into the registry system. I'm not a huge fan of putting resource packs into a util class when util classes are meant to represent utilities rather than data storage/management.

Redned235 commented on core/src/main/java/org/geysermc/geyser/network/UpstreamPacketHandler.java:
```diff
@@ -68,10 +70,13 @@
-    private Deque<String> packsToSent = new ArrayDeque<>();
+    private final Deque<String> packsToSent = new ArrayDeque<>();
+
+    private PlayerResourcePackLoadEvent resourcePackLoadEvent;
```
Any reason this is an instance variable?

onebeastchris replied:
The packs of the event are also needed when the client requests a pack; therefore I'm storing them here for this instance.. unless there's some better way to get those packs from this fired event

Redned235 commented on api/build.gradle.kts:
```diff
@@ -4,4 +4,7 @@ plugins {
 dependencies {
     api(libs.base.api)
+    // https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-annotations
+    implementation("com.fasterxml.jackson.core:jackson-annotations:2.14.0")
```
This should not be necessary once all the other suggestions are applied.

Creeperz653 commented:
are you saying we can have per server packs?

onebeastchris commented:
Yup! This PR would give the framework for it, so e.g. an extension or a plugin could be made for it.
As for camotoys/redneds reviews; Thank you and I'll try to address those quickly

Creeperz653 commented:
> Yup! This PR would give the framework for it, so e.g. an extension or a plugin could be made for it.
> As for camotoys/redneds reviews; Thank you and I'll try to address those quickly 

Sick but y'all said it wasn't possible

Jens-Co commented:
> > Yup! This PR would give the framework for it, so e.g. an extension or a plugin could be made for it.
> > As for camotoys/redneds reviews; Thank you and I'll try to address those quickly
> 
> Sick but y'all said it wasn't possible

We never said it was not possible, it's just not something we would be actively working on as there are other/more important things being worked on.

onebeastchris commented:
+ it's still not seamless pack switching/applying; you have to disconnect+reconnect the player (either manually, or with a transfer packet) to change which packs are applied. But since this could be done automatically upon e.g. server switching.. close enough I'd say

onebeastchris commented:
Not yet done; need to figure out how to properly load the packs now ~ should get that fixed fairly quickly

YHDiamond commented on api/src/main/java/org/geysermc/geyser/api/packs/GeyserResourcePackManifest.java:
```diff
@@ -0,0 +1,241 @@
+                    .replace('[', ' ')
+                    .replaceAll(" ", "").split(",");
+
+            return new Version() {
+                @Override
+                public int major() {
+                    return Integer.parseInt(split[0]);
+                }
+
+                @Override
+                public int minor() {
+                    return Integer.parseInt(split[1]);
+                }
+
+                @Override
+                public int patch() {
+                    return Integer.parseInt(split[2]);
+                }
+            };
```
I think this could be replaced with this, less repeated code.
```suggestion
            return fromArray(new int[]{Integer.parseInt(split[0]), Integer.parseInt(split[1]), Integer.parseInt(split[2]));   
```

Konicai commented on core/build.gradle.kts:
```diff
@@ -66,6 +66,10 @@ configurations.api {
+java {
+    sourceCompatibility = JavaVersion.VERSION_16
+    targetCompatibility = JavaVersion.VERSION_16
+}
```
Was this added because of the current indra issue? This should be removed

onebeastchris replied:
yeee i'll yeet it

Konicai commented on core/src/main/java/org/geysermc/geyser/registry/loader/ResourcePackLoader.java:
```diff
@@ -0,0 +1,157 @@
+            resourcePacks = Files.walk(directory)
+                    .filter(PACK_MATCHER::matches)
+                    .collect(Collectors.toCollection(ArrayList::new)); // toList() does not guarantee mutability
+        } catch (Exception e) {
+            GeyserImpl.getInstance().getLogger().error("Could not list packs directory", e);
+            return packMap;
+        }
+
+        GeyserLoadResourcePacksEvent event = new GeyserLoadResourcePacksEvent(resourcePacks);
+        GeyserImpl.getInstance().eventBus().fire(event);
```
RFC: Do we still fire the event if there was an exception listing the packs directory, as listeners will still be able to add to the list?

Redned235 replied:
Updated so event is still fired.

Konicai commented on core/src/main/java/org/geysermc/geyser/registry/loader/ResourcePackLoader.java:
```diff
@@ -0,0 +1,157 @@
+        GeyserLoadResourcePacksEvent event = new GeyserLoadResourcePacksEvent(resourcePacks);
+        GeyserImpl.getInstance().eventBus().fire(event);
+
+        for (Path path : event.resourcePacks()) {
+            try {
+                GeyserResourcePack pack = readPack(path);
+                packMap.put(pack.manifest().header().uuid().toString(), pack);
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
```
API change: Paths that do **not** end in `.zip` or `.mcpack` are no longer filtered from the event.

Redned235 replied:
Updated.

Konicai commented on core/src/main/java/org/geysermc/geyser/registry/loader/ResourcePackLoader.java:
```diff
@@ -0,0 +1,157 @@
+                // Outdated naming?
+                manifest = readManifest(zip, "pack_manifest.json");
+                if (manifest == null || manifest.header().uuid() == null) {
+                    throw new IllegalArgumentException(path.getFileName() + " does not contain a valid pack_manifest.json or manifest.json");
```
Kind of ugly because this will be caught and rethrown below... not sure if it matters that much.

Konicai reviewed (commented):
Some thoughts I had while writing, or important changes I wanted to point out

onebeastchris commented:
no longer a draft, ready for reviews:
- api changes (including pack loading) work, tested them with the same extension (but updated to latest changes)
- pack loading (encrypted packs aswell according to @Konicai ) works same as before

Changes since last reviews:
- ResourcePack and ResourcePackManifest are now interfaces, with records implementing them in geyser core
- huge thanks to konica to re-working pack loading system & exposing pack loading properly to api
- (hopefully) proper documentation of new api
- Packs are loaded via Registry now

Konicai commented on core/src/main/java/org/geysermc/geyser/pack/GeyserResourcePackManifest.java:
```diff
@@ -27,16 +27,22 @@
 import java.util.UUID;
 
-public record GeyserResourcePackManifest(Header header) implements ResourcePackManifest {
+    public record GeyserResourcePackManifest(Integer formatVersion, Header header, Collection<Module> modules, Collection<Dependency> dependencies) implements ResourcePackManifest {
```
Fix indentation

Konicai commented on core/src/main/java/org/geysermc/geyser/pack/GeyserResourcePackManifest.java:
```diff
@@ -27,16 +27,22 @@
+    public record GeyserResourcePackManifest(Integer formatVersion, Header header, Collection<Module> modules, Collection<Dependency> dependencies) implements ResourcePackManifest {
 
-    public record Header(UUID uuid, int[] version) implements ResourcePackManifest.Header {
+        public record Header(UUID uuid, int[] version, String name, String description, String versionString, int[] minimumSupportedMinecraftVersion) implements ResourcePackManifest.Header {
```
Remove the `versionString` member from
the Header record

onebeastchris replied:
aka just leave the override?

Konicai replied:
Yeah, versionString() is just a utility method, it doesn't exist in the manifest

Konicai commented:
Also, you are missing jackson annotations on the `formatVersion` and `minimumSupportedMinecraftVersion`. Their keys in the manifest are actually `format_version` and `min_engine_version` (which is why the annotations are required). Look at the old deleted ResourcePack class for how it was done

I don't think jackson complains about missing members unless you tell it to, so you'll have to test to make sure that all the fields are being properly loaded

Konicai commented on core/src/main/java/org/geysermc/geyser/registry/loader/ResourcePackLoader.java:
```diff
@@ -74,8 +76,8 @@ public Map<String, ResourcePack> load(Path directory) {
-            resourcePacks = Files.walk(directory)
+        try (Stream<Path> stream = Files.walk(directory)) {
+            resourcePacks = stream
                     .filter(PACK_MATCHER::matches)
```
put `.filter(...` on the same line as `stream`

onebeastchris commented on core/src/main/java/org/geysermc/geyser/registry/loader/ResourcePackLoader.java:
```diff
@@ -106,19 +108,31 @@ public Map<String, ResourcePack> load(Path directory) {
+                            + ", " + name.length() + " characters long). This will cause problems on some Bedrock platforms." +
+                            " Please rename it to be shorter, or reduce the amount of folders needed to get to the file.");
+                }
+                if (name.contains("manifest.json")) {
+                    try {
+                        manifestReference.set(FileUtils.loadJson(zip.getInputStream(x), GeyserResourcePackManifest.class));
```
![image](https://github.com/GeyserMC/Geyser/assets/105284508/baf74eeb-834d-4e83-9d1a-8f7268bc91f5)
could be handled like this if desired

Konicai replied:
If a valid file is found first, and then an invalid file is found afterwards, that would overwrite the valid file

just don't call `set()` if the uuid in the header is null

onebeastchris replied:
re-implemented the check

Konicai commented on core/src/main/java/org/geysermc/geyser/registry/loader/ResourcePackLoader.java:
```diff
@@ -0,0 +1,154 @@
+     */
+    public static GeyserResourcePack readPack(Path path) throws IllegalArgumentException {
+        AtomicReference<GeyserResourcePackManifest> manifestReference = new AtomicReference<>();
+        GeyserResourcePack pack;
```
Remove this variable as the pack is returned in only one location. Apart from that everything seems alright to me

The ResourcePack/Header/Manifest API and what it will look like in the future just needs to be considered by Redned

onebeastchris replied:
resolved; thanks

Konicai commented on core/src/main/java/org/geysermc/geyser/pack/path/GeyserPathPackCodec.java:
```diff
@@ -0,0 +1,74 @@
+        return this.path;
+    }
+
+    @Override
+    public byte @NonNull [] sha256() {
+        return FileUtils.calculateSHA256(this.path);
+    }
+
+    @Override
+    public long size() {
+        try {
+            return Files.size(this.path);
+        } catch (IOException e) {
+            throw new RuntimeException("Could not get file size of path " + this.path, e);
+        }
+    }
```
These are both called many times when sending pack info, and size is called each time a chunk is sent - would it be worth it to cache these? I have no idea what the time complexity is like but it sounds like something that you would want to avoid for large packs and a lot of players.

It's possible the file is changed afterwards and the cached hash/size becomes inaccurate. When the hash/size is calculated, we could store the lastModifiedTime, and check if it changed whenever sha256() or size() is called. Doesn't sound beautiful but it's simple.

Redned235 replied:
Addressed

Konicai commented on api/src/main/java/org/geysermc/geyser/api/pack/PackCodec.java:
```diff
@@ -0,0 +1,83 @@
+ * Represents a pack codec that can be used
+ * to provide resource packs to clients.
+ */
+public abstract class PackCodec {
```
Could just be an interface? Same with PathPackCodec

Redned235 replied:
Has a protected abstract method so that’s not possible unfortunately.

Konicai replied:
Just so that people have to call it on ResourcePack? I'm not saying the method shouldn't be in ResourcePack, but all it does is directly call `create()`

Redned235 replied:
We could expose it that way, but it'd mean there's duplicate API. We could also go the route of not having `create` in ResourcePack, but it's definitely far nicer than doing `PackCodec.path(Paths.get(...)).create()`

Konicai replied:
Fair. And I guess keeping it abstract would allow use to add state/behaviour in the future that we don't want to expose publicly

Konicai commented on api/src/main/java/org/geysermc/geyser/api/pack/ResourcePackManifest.java:
```diff
@@ -0,0 +1,180 @@
+     *
+     * @return the format version
+     */
+    @NonNull
+    Integer formatVersion();
```
NonNull but boxed? Also applies to GeyserResourcePackManifest

Tim203 commented on core/src/main/java/org/geysermc/geyser/registry/loader/ResourcePackLoader.java:
```diff
@@ -0,0 +1,150 @@
+        List<Path> resourcePacks;
+        try (Stream<Path> stream = Files.walk(directory)) {
+            resourcePacks = stream.filter(PACK_MATCHER::matches)
+                    .collect(Collectors.toCollection(ArrayList::new)); // toList() does not guarantee mutability
```
We don't need mutability right? And otherwise `.collect(Collectors.toList())` should be mutable as well.

Konicai replied:
[Collectors#toList - There are no guarantees on the type, mutability, serializability, or thread-safety of the List returned](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collectors.html#toList--)

Needs to be mutable because its passed to GeyserLoadResourcePacksEvent

Tim203 reviewed (changes requested):
These are mostly related to the API part, as those are harder to change once they're merged.
And I also agree with the comments of Konicai.

Konicai commented on api/src/main/java/org/geysermc/geyser/api/event/bedrock/SessionLoadResourcePacksEvent.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Called when Geyser initializes a session for a new Bedrock client and is in the process of sending resource packs.
+ */
+public class SessionLoadResourcePacksEvent extends ConnectionEvent {
```
latter seems good

Redned235 replied:
We’ve adopted the session naming already for other events.

Konicai commented on api/src/main/java/org/geysermc/geyser/api/pack/PackCodec.java:
```diff
@@ -0,0 +1,83 @@
+     *
+     * @return the hash of the resource pack
+     */
+    public abstract byte @NonNull [] sha256();
```
UpstreamPacketHandler directly uses the byte[] so making it a string would just be intermediary

Redned235 replied:
The nonnull in the way it is here means the array itself cannot be null. Having it before the byte part signifies the byte itself is null, and is flagged as a problem in IntelliJ bc you can’t have a null primitive.

Redned235 commented on api/src/main/java/org/geysermc/geyser/api/GeyserApi.java:
```diff
@@ -107,6 +108,22 @@ public interface GeyserApi extends GeyserApiBase {
+     * @return the path to the Geyser config directory
+     */
+    @NonNull
+    Path configDirectory();
```
Don't really see a reason not to.

onebeastchris replied:
I would keep it, never hurts to expose useful bits and pieces in api

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/event/bedrock/SessionLoadResourcePacksEvent.java:
```diff
@@ -0,0 +1,55 @@
+     * The keys are the pack ID and the values are the resource packs themselves.
+     */
+    @NonNull
+    public Map<String, ResourcePack> packs() {
```
I could add such a method, sure, but I don't really see a need for it if we're going to expose an unmodifiable list of packs instead.. since the extension/api user could iterate over it to get a pack if so desired

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/pack/ResourcePackManifest.java:
```diff
@@ -0,0 +1,180 @@
+         * @return the UUID
+         */
+        @NonNull
+        UUID uuid();
```
since the map is no longer exposed, i dont think the change is necessary - the reason for why the packid is stored as a string is for ease of use - the ResourcePackClientResponse packet sends a list of strings
![image](https://github.com/GeyserMC/Geyser/assets/105284508/443e5f2b-9fdb-498e-b7f1-fefc20135e6b)

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/pack/ResourcePackManifest.java:
```diff
@@ -0,0 +1,180 @@
+         *
+         * @return the version
+         */
+        int @NonNull [] version();
```
I had to keep the int[] array due to jackson complaining about serializing the array into seemingly anything else - however, i did change the versionString() method - if that isnt quite what is wished for, i wouldnt know how to make jackson parse the array into anything.. even a custom constructor for a version object that takes in an array doesnt work

Redned235 replied:
Version interface would be cleaner. See [this](https://www.baeldung.com/jackson-custom-serialization) for how to handle custom serialization.

onebeastchris replied:
Thanks for the docs; resolved.

onebeastchris commented on core/src/main/java/org/geysermc/geyser/registry/DeferredRegistry.java:
```diff
@@ -0,0 +1,180 @@
+ *
+ * @param <M> the value being held by the registry
+ */
+public class DeferredRegistry<M> implements IRegistry<M> {
```
Cant really comment on this, maybe @Redned235 can clarify?

Redned235 replied:
Set this to final. I don't envision any instance where we'd need to extend this, but if the need ever arises in the future, we can deal with it then.

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/pack/ResourcePackManifest.java:
```diff
@@ -0,0 +1,210 @@
+         * @return the version string
+         */
+        @NonNull
+        String versionString();
```
I could leave it here, or move it to version.. definitely needed though

onebeastchris replied:
Moved to Version interface, and renamed to toString() since version().versionString() would be kinda ugly

Creeperz653 commented:
Woooooo
